In [30]:
import pandas as pd
import numpy as np
import re

In [47]:
def parse_time_str(s: str) -> list[float]:
    pattern = r":(\d+\.?\d*)"
    numbers = re.findall(pattern, s)
    # convert to float
    result = [float(num) for num in numbers]
    return result
def parse_time_series(s):
    # convert to float
    return np.array(parse_time_str(s), dtype=np.float64)

In [20]:
df = pd.read_csv('result.csv')
df.head()

name                                 repository          dirname  \
0     socket2       https://github.com/rust-lang/socket2          socket2   
1      flate2     https://github.com/rust-lang/flate2-rs        flate2-rs   
2  libloading  https://github.com/nagisa/rust_libloading  rust_libloading   
3     openssl   https://github.com/sfackler/rust-openssl     rust-openssl   
4   zstd-safe          https://github.com/gyscos/zstd-rs          zstd-rs   

   valid_proj  build_success  \
0       False           True   
1       False           True   
2        True           True   
3       False          False   
4        True           True   

                                   normal_build_time  ffi_checker_success  \
0   real_time:1.12s, user_time:1.03s, sys_time:0.40s                 True   
1   real_time:0.87s, user_time:1.28s, sys_time:0.37s                 True   
2   real_time:0.31s, user_time:0.20s, sys_time:0.17s                 True   
3                                                NaN                False   
4  real_time:6.30s, user_time:12.50s, sys_time:2.50s                 True   

                              ffi_checker_build_time  \
0   real_time:1.72s, user_time:1.29s, sys_time:0.49s   
1   real_time:1.27s, user_time:1.45s, sys_time:0.43s   
2   real_time:0.52s, user_time:0.32s, sys_time:0.19s   
3                                                NaN   
4  real_time:9.00s, user_time:13.50s, sys_time:2.97s   

   ffi_checker_analysis_time  
0                        NaN  
1                        NaN  
2                        NaN  
3                        NaN  
4                        NaN

In [21]:
build_success_cnt = df[df["build_success"] == True]['build_success'].sum()
build_success_cnt

np.int64(124)

In [22]:
ffi_success_cnt = df[df["ffi_checker_success"] == True]['ffi_checker_success'].sum()
ffi_success_cnt

np.int64(124)

In [23]:
valid_cnt = df[df["valid_proj"] == True]['valid_proj'].sum()
valid_cnt

np.int64(28)

In [51]:
normal_build_times = df[df['normal_build_time'].notna()]["normal_build_time"].apply(parse_time_series)
build_time_sum = normal_build_times.sum(axis=0)
build_time_sum

array([1617.83, 4074.86,  738.88])

In [52]:
ffi_checker_build_times = df[df['ffi_checker_build_time'].notna()]["ffi_checker_build_time"].apply(parse_time_series)
gen_ir_time_sum = ffi_checker_build_times.sum(axis=0)
gen_ir_time_sum

array([2228.09, 5212.66,  944.14])

In [53]:
(gen_ir_time_sum - build_time_sum) / build_time_sum * 100

array([37.72089775, 27.92243169, 27.77988307])

使用FFI Checker功能额外带来的编译时间开销为 real_time: 37.72% user_time: 27.92% sys_time: 27.78%